### PR TITLE
gstvkvideoparser: rework the parser to be reloadable as a library in both windows and linux environment

### PR DESCRIPTION
**Pull request by dabrain34**
Due to glib limitations with GType over a process lifetime, the library can not be unloaded easily.
On linux the problem was solved by using the -z,nodelete linker option but on windows the problem was still remaining as the windows linker/system does not support this option.

- the VK parser/decoder (H264 H265) elements have been moved into a plugin which is loaded by the registry.
- The GstVkVideoParser GObject has been tranformed to a C++ object to get rid of the GType
- The gst_harness has been imported to remove the GstTestClock use avoiding the GType to be registered.
- A dlopen test has been created to reproduce the CTS issue.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -38,7 +38,9 @@ jobs:
           TERM: dumb
 
       - name: Test
-        run: meson test --verbose -C builddir
+        run:
+          export GST_PLUGIN_PATH=`pwd`/builddir/plugins;
+          meson test --verbose -C builddir
```
Comment (ceyusa) [inline .github/workflows/main.yml:43]:
```suggestion
        run:
          meson test --verbose -C builddir
        env:
          GST_PLUGIN_PATH: builddir/plugins;

```

File: lib/gstvkvideoparser.cpp [hunk 2]
```diff
@@ -89,49 +92,44 @@ process_messages (GstVkVideoParser * self)
   }
 }
 
-static void
-gst_vk_video_parser_dispose (GObject * object)
+void GstVkVideoParser::Dispose ()
```
Comment (ceyusa) [inline lib/gstvkvideoparser.cpp:95]: What about using a C++ destructor ~GstVkVideoParser() ?

File: lib/gstvkvideoparser.cpp [hunk 3]
```diff
@@ -89,49 +92,44 @@ process_messages (GstVkVideoParser * self)
...
 
-  gst_object_unref (self->bus);
-
-  G_OBJECT_CLASS (gst_vk_video_parser_parent_class)->dispose (object);
+  gst_object_unref (this->m_bus);
 }
 
-static void
-gst_vk_video_parser_constructed (GObject * object)
+bool GstVkVideoParser::Build ()
```
Comment (ceyusa) [inline lib/gstvkvideoparser.cpp:111]: What about using the C++ constructor for this?

File: lib/gstvkvideoparser.cpp [hunk 4]
```diff
@@ -142,125 +140,59 @@ gst_vk_video_parser_constructed (GObject * object)
...
-}
 
-GstFlowReturn
-gst_vk_video_parser_push_buffer (GstVkVideoParser * self, GstBuffer * buffer)
+GstFlowReturn GstVkVideoParser::PushBuffer (GstBuffer * buffer)
 {
   GstFlowReturn ret;
 
-  GST_DEBUG_OBJECT (self, "Pushing buffer: %" GST_PTR_FORMAT, buffer);
+  GST_DEBUG_OBJECT (this, "Pushing buffer: %" GST_PTR_FORMAT, buffer);
```
Comment (ceyusa) [inline lib/gstvkvideoparser.cpp:174]: ditto

File: lib/gstvkvideoparser.cpp [hunk 5]
```diff
@@ -142,125 +140,59 @@ gst_vk_video_parser_constructed (GObject * object)
...
 
-  ret = gst_harness_push (self->parser, buffer);
+  ret = gst_harness_push (m_parser, buffer);
   if (ret != GST_FLOW_OK && ret != GST_FLOW_EOS) {
-    GST_WARNING_OBJECT (self, "Couldn't push buffer: %s",
+    GST_WARNING_OBJECT (this, "Couldn't push buffer: %s",
```
Comment (ceyusa) [inline lib/gstvkvideoparser.cpp:178]: ditto

File: lib/gstvkvideoparser.cpp (hunk 3 above)
Comment (dabrain34) [reply lib/gstvkvideoparser.cpp:111]: To check in GstVkVideoDecoderParser::Initialize that the initialization succeeded

File: lib/gstvkvideoparser.cpp [hunk 6]
```diff
@@ -142,125 +140,59 @@ gst_vk_video_parser_constructed (GObject * object)
...
 
   return ret;
 }
 
-GstFlowReturn
-gst_vk_video_parser_eos (GstVkVideoParser * self)
+GstFlowReturn GstVkVideoParser::Eos ()
 {
-  GST_DEBUG_OBJECT (self, "Pushing EOS");
+  GST_DEBUG_OBJECT (this, "Pushing EOS");
```
Comment (ceyusa) [inline lib/gstvkvideoparser.cpp:190]: ditto

File: lib/meson.build [hunk 7]
```diff
@@ -1,56 +1,85 @@
 pkgconfig = import('pkgconfig')
 
-sources = files(
-  'vkvideodecodeparser.cpp',
-  'gstvkh264dec.cpp',
-  'gstvkh265dec.cpp',
-  'gstvkvideoparser.cpp',
-  'videoutils.c',
+codecparser_sources = files (
```
Comment (ceyusa) [inline lib/meson.build:3]: isn't possible to bundle gstharness with gstvkvideoparser avoiding another static library?

Comment (ceyusa) [reply lib/meson.build:3]: doh, it is bundled... didn't read it right

**Review by ceyusa — APPROVED** [review]
it's a huge changeset... I'm not sure about the design, but it's what we have right now

why you didn't remove the gsttestclock in harndes?

File: lib/meson.build (hunk 7 above)
Comment (dabrain34) [reply lib/meson.build:3]: I wanted to stay the closest from the orginal code but just setting the testclock to NULL and not invoke `GST_TEST_CLOCK_CAST (gst_test_clock_new ());. `

**Pull request merged**